ddl/column.go

```diff
@@ -633,36 +640,41 @@ func onSetDefaultValue(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	return updateColumnDefaultValue(t, job, newCol, &newCol.Name)
 }
 
-func (w *worker) onModifyColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
-	newCol := &model.ColumnInfo{}
-	oldColName := &model.CIStr{}
+func needChangeColumnData(oldCol, newCol *model.ColumnInfo) bool {
+	toUnsigned := mysql.HasUnsignedFlag(newCol.Flag)
+	originUnsigned := mysql.HasUnsignedFlag(oldCol.Flag)
+	if newCol.Flen > 0 && newCol.Flen < oldCol.Flen || toUnsigned != originUnsigned {
+		return true
+	}
+
+	return false
+}
+
+func (w *worker) onModifyColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	var (
+		newCol                *model.ColumnInfo
+		oldColName            *model.CIStr
+		modifyColumnTp        byte
+		updatedAutoRandomBits uint64
+		changingCol           *model.ColumnInfo
+		changingIdxs          []*model.IndexInfo
+	)
 	pos := &ast.ColumnPosition{}
-	var modifyColumnTp byte
-	var updatedAutoRandomBits uint64
-	err := job.DecodeArgs(newCol, oldColName, pos, &modifyColumnTp, &updatedAutoRandomBits)
+	err := job.DecodeArgs(&newCol, &oldColName, pos, &modifyColumnTp, &updatedAutoRandomBits, &changingCol, &changingIdxs)
```

Comment:
The last two arguments are always decoded to nil. Are they still in TODO?

Reply:
Which two parameters? Only when the job is none state, `changingCol` and `changingIdxs` are nil.